Map block: Record a Tracks event on Map load on WordPress.com

Copons opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Record a Tracks event every time the Map block is loaded on WordPress.com and Atomic.
The only event property is the site ID.

The reason is to be able to monitor the usage of the Mapbox API in case of abuses or excesses, and to identify and blacklist any possible problematic sites.

No events are tracked on Jetpack (non-Simple, non-Atomic) sites.

#### Re: GDPR review

This event is recorded both on the editor and on the front end.
We have no interest in knowing the user, but only the site ID.
I would have liked to use `tracks_record_event` without providing a user, but as far as I can see it's not possible to, say, pass `null`, as the tracking system [requires a `WP_User` object](https://github.com/Automattic/jetpack/blob/be32f2677a02b4f4e4e84146aa8dcf3bbf97c08b/packages/tracking/src/class-tracking.php#L132).
For this reason, I've decided to pass `wp_get_current_user()`, and let the system decide what to do with it.

I wonder if it makes sense to simply pass a `new WP_User()` instead.

Either way, I'd love to make sure I'm not messing up with GDPR in any ways whatsoever.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Enhancement

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

To test this in a timely manner, please add an `error_log` to both event recordings. On record success, it should just log `true`.
E.g.
```php
$t = tracks_record_event( $user, $event_name, array( 'site_id' => get_current_blog_id() ) );
error_log( json_encode( $t ) );
```

Jetpack Site

* Open the editor, add a Map block, and insert a personal Mapbox access token.
* Save the post, and view it on the front end.
* Check the error log and make sure there are no logs caused by the event tracking.

Simple Site

* Checkout the twin diff and manually add the `error_log`s.
* Open the editor, add a Map block.
* Save the post, and view it on the front end.
* Check the error log and make sure there are 2 logs: 1 for the editor and 1 for the front end.
* Now add another Map blocks to the post.
* Save and open on the front end.
* Check the error log: make sure there are 4 new logs: 1 for each Map blocks in the editor and in the front end.

Atomic Site

* Install and activate the [Jetpack Beta](https://github.com/Automattic/jetpack-beta/releases/latest) plugin.
* In wp-admin, navigate to Jetpack -> Jetpack Beta; search for the `update/map-block-tracks` feature branch, and activate it.
* I'm not sure how to check the error log on Atomic, so head to the wp-admin plugin editor and add some `var_dump`s instead. Make sure you're modifying the correct Jetpack (one is the normal version, the other contains this very branch).
* Open the editor, add a Map block.
* Save the post, and view it on the front end.
* Make sure there are 2 var dumps: 1 for the editor and 1 for the front end.
* Now add another Map blocks to the post.
* Save and open on the front end.
* Make sure there are 4 var dumps: 1 for each Map blocks in the editor and in the front end.

Automatticians

After a while (10+ minutes), search for `jetpack_map_block_mapbox_load` and check if the results are consistent with your tests.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

Copons committed:
Record a Tracks event on Map load

jeherve commented on extensions/blocks/map/map.php:
```diff
@@ -40,6 +40,22 @@ function jetpack_get_mapbox_api_key() {
+		return;
+	} elseif ( jetpack_is_atomic_site() ) {
+		$tracking = new Automattic\Jetpack\Tracking();
+		$tracking->tracks_record_event( $user, $event_name, array( 'site_id' => Jetpack_Options::get_option( 'id' ) ) );
```
Maybe you can use `record_user_event` here in Jetpack? It offers the advantage of taking care of adding the blog ID (and more) for you:
https://github.com/Automattic/jetpack/blob/1b981c3573c82a4c83b3c212a241b59097691792/packages/tracking/src/class-tracking.php#L60-L77

Copons replied:
Since we need the site ID to do usage checks, I'd rather explicitly add it as custom event property and have full control over it.
It's obviously very unlikely, but what if at some point we decide to stop recording the `blog_id`, or change its name?

(Also, but this might very well be my fault: on Tracks the global property is called `blogid` and I've seen that it often returns `NULL`... Better safe than sorry I guess?)

jeherve replied:
> Since we need the site ID to do usage checks, I'd rather explicitly add it as custom event property and have full control over it.

Since you're pulling it from the same place as `record_user_event` does, you should always have similar results.

> what if at some point we decide to stop recording the blog_id, or change its name?

That's unlikely, as it's a key piece on both Jetpack and wpcom.

If you want to be double sure you could add an extra `site_id` property to the data you pass to `record_user_event`, so you'd have it twice if you'd like. It does seem a little bit overkill though.

> this might very well be my fault: on Tracks the global property is called blogid and I've seen that it often returns NULL... Better safe than sorry I guess?

That is odd indeed, and may be worth its own investigation.

Copons replied:
Might very well be because people are using `tracks_record_event` instead of `record_user_event` 😄

jeherve commented on extensions/blocks/map/map.php:
```diff
@@ -126,6 +144,8 @@ function jetpack_map_block_render_single_block_page() {
 
 	add_filter( 'jetpack_is_amp_request', '__return_false' );
 
+	jetpack_record_mapbox_load_event();
```
I am not sure I understand why we call this here again. Am I missing something?

Copons replied:
This was slightly confusing to me too, and I hope I got it right (cc @jeffersonrabb).

The Map block's AMP implementation is peculiar, or at least it's peculiar to me with no familiarity with AMP.

When loaded normally (not in AMP), this file is a simple server-side rendering of the Map block, similar to any other.

But it also has a function hooked to the `wp` action (that runs very early in the execution) that, if the URL has two Map block specific query params, instead of rendering the site, it renders **only** the map specified by the query params.
You can test this yourself with an URL like this:
`https://www.example.org?map-block-post-id={ id of a post containing Map blocks }&map-block-counter={ number of the Map block to render }`

Now, when we load the block in AMP, instead of normally rendering the Map block, we render an AMP iframe using that kind of URL as source:
`<amp-iframe src="https://example.org?map-block-post-id=123&map-block-counter=1` />`
In other words, instead of directly loading the map via JS, we load it into an iframe (to make AMP happy, I guess).

Because of this kind of switcheroo (and some early returns that would affect the tracking accuracy), I personally feel more confident in tracking the Mapbox load _as late as possible_, which is, right next to `Jetpack_Gutenberg::load_assets_as_required( 'map' );`. 😄

jeherve reviewed (changes requested):
> I would have liked to use tracks_record_event without providing a user, but as far as I can see it's not possible to, say, pass null, as the tracking system requires a WP_User object.
> For this reason, I've decided to pass wp_get_current_user(), and let the system decide what to do with it.
>
> I wonder if it makes sense to simply pass a new WP_User() instead.

In the past we had opted to fill in blank user submissions with the blog owner's ID. Here is an example:
https://github.com/Automattic/jetpack/blob/6fec7802eef12374cbbd3c97b5f1def5c672034d/modules/contact-form/grunion-contact-form.php#L3578-L3604

As you can see we had also prefixed the event name differently whether the call was from a WordPress.com or a Jetpack site. I don't know if that's something you are interested in?

Copons commented on extensions/blocks/map/map.php:
```diff
@@ -40,6 +40,22 @@ function jetpack_get_mapbox_api_key() {
+	if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+		require_lib( 'tracks/client' );
+		tracks_record_event( $user, $event_name, array( 'site_id' => get_current_blog_id() ) );
+		return;
```
Nope, that's a leftover from an earlier implementation! 🙂

Copons commented on extensions/blocks/map/map.php:
```diff
@@ -40,6 +40,22 @@ function jetpack_get_mapbox_api_key() {
+		require_lib( 'tracks/client' );
+		tracks_record_event( $user, $event_name, array( 'site_id' => get_current_blog_id() ) );
+		return;
+	} elseif ( jetpack_is_atomic_site() ) {
```
I'm not opposed to the change obviously, but I'd love to have more context on this.
As far as I can tell, `is_active` checks if JP is connected to WPCOM.
So, is it possible for an Atomic site to disconnect from JP?

Or maybe you mean that `is_active` also checks other stuff (e.g. if a site has accepted the ToS). 🤔

jeherve replied:
> As far as I can tell, is_active checks if JP is connected to WPCOM.

That's correct.

> is it possible for an Atomic site to disconnect from JP?

Normally not, but there are some use-cases where the connection breaks.

It may be enough of an edge-case for us to ignore, though. I am not insisting on it.

Copons commented on extensions/blocks/map/map.php:
```diff
@@ -79,6 +95,8 @@ function jetpack_map_block_load_assets( $attr, $content ) {
 		);
 	}
 
+	jetpack_record_mapbox_load_event();
```
See https://github.com/Automattic/jetpack/pull/14465#discussion_r371868039

Copons committed:
Improve the tracking on WPCOM

Copons committed:
Remove site_id from the event custom properties

Copons commented:
@jeherve I've updated this following your suggestions.
I've also removed the explicit tracking of the site ID, relying on the implicit one.

I've tested this on a Simple site, and the `blogid` is tracked as expected.
I've also tested this on an Atomic site with the beta plugin, but the events aren't showing up in the Tracks Live View (after 20+ minutes), even though the record function is called and returns `true`, so I can't be 100% about them.

jeffersonrabb commented on extensions/blocks/map/map.php:
```diff
@@ -79,6 +93,8 @@ function jetpack_map_block_load_assets( $attr, $content ) {
 		);
 	}
 
+	jetpack_record_mapbox_wpcom_load_event();
```
I believe this will double-count Map blocks in AMP requests. Here's what I'd recommend, which will cause the tracking function to fire once per Map block regardless of AMP status:

1) In `jetpack_record_mapbox_wpcom_load_event()`, bail if the AMP query params are present, like so:

```
function jetpack_record_mapbox_wpcom_load_event() {

	$map_block_counter = isset( $_GET, $_GET['map-block-counter'] ) ? absint( $_GET['map-block-counter'] ) : null;
	// phpcs:ignore WordPress.Security.NonceVerification
	$map_block_post_id = isset( $_GET, $_GET['map-block-post-id'] ) ? absint( $_GET['map-block-post-id'] ) : null;

	if ( $map_block_counter && $map_block_post_id ) {
		return;
	}
```

2) In `jetpack_map_block_load_assets()` move `jetpack_record_mapbox_wpcom_load_event();` above the AMP check.

3) Remove entirely the second call to `jetpack_record_mapbox_wpcom_load_event()`.

Copons replied:
@jeffersonrabb AFAICS, in AMP requests, we wouldn't get to this: https://github.com/Automattic/jetpack/blob/8f0098b3e3e3beb622c1f37bb7df1cb6f381dd29/extensions/blocks/map/map.php#L96
Because we early return here: https://github.com/Automattic/jetpack/blob/8f0098b3e3e3beb622c1f37bb7df1cb6f381dd29/extensions/blocks/map/map.php#L87-L92

But yeah, reading again the code, I guess that moving the recording above the AMP check  (and removing the second call) should be safe enough in all cases.

Copons committed:
Simplify the even recording

creativecoder commented on extensions/blocks/map/map.php:
```diff
@@ -40,6 +40,20 @@ function jetpack_get_mapbox_api_key() {
+ * Record a Tracks event every time the Map block is loaded on WordPress.com and Atomic.
+ */
+function jetpack_record_mapbox_wpcom_load_event() {
+	$event_name = 'wpcom_map_block_mapbox_load';
```
I believe prefixing the event with `wpcom` implies the event is for Simple Sites (on wpcom). I don't think we need to track `wpcom` and `jetpack` events separately, it's probably more convenient to keep this as a single event.

What about simply `map_block_mapbox_load`?

creativecoder replied:
Or maybe `map_block_mapbox_wpcom_key_load`?

Copons replied:
Yeah I was on the fence about this event name. 🤔 
I guess `map_block_mapbox_wpcom_key_load` is the most descriptive and obvious.

> I don't think we need to track wpcom and jetpack events separately

Agreed. I don't see much value in knowing the site type for this event.

creativecoder commented on extensions/blocks/map/map.php:
```diff
@@ -49,6 +63,8 @@ function jetpack_get_mapbox_api_key() {
  * @return string
  */
 function jetpack_map_block_load_assets( $attr, $content ) {
+	jetpack_record_mapbox_wpcom_load_event();
```
If the goal is to "monitor the usage of the Mapbox API in case of abuses or excesses, and to identify and blacklist any possible problematic sites." (from the PR description)... shouldn't we only trigger this event if a WordPress.com key is being used?

Copons replied:
This is a very good point! Not sure how I've overlooked it 🤦‍♂

Copons committed:
Only track the event if the site uses the WPCOM access token

jeherve reviewed (approved):
> I've tested this on a Simple site, and the blogid is tracked as expected.
I've also tested this on an Atomic site with the beta plugin, but the events aren't showing up in the Tracks Live View (after 20+ minutes), even though the record function is called and returns true, so I can't be 100% about them.

This seems to work for me in Jetpack. I could not get the events to show when testing on a simple site though.

I'll let others give this a try as well, but the code looks good for me.

Copons commented:
Note: this is dependent on https://github.com/Automattic/jetpack/pull/14597

jeherve reviewed (changes requested):
This now needs a rebase.

Copons commented:
Replaced by #14779 (it was simpler than rebasing 😄)
